2.train-VAE/betatcvae.py

```diff
@@ -0,0 +1,300 @@
+# betatcvae.py
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from typing import List
+import pandas as pd 
+import numpy as np
+import pathlib
+import math
+
+
+class BetaTCVAE(nn.Module):
+    """
+    Beta Total Correlation Variational Autoencoder (Beta-TCVAE) implementation.
+
+    Attributes:
+    -----------
+    latent_dim : int
+        Dimensionality of the latent space.
+    anneal_steps : int
+        Number of steps to anneal the KL divergence term.
+    alpha : float
+        Weight of the mutual information term in the loss function.
+    beta : float
+        Weight of the total correlation term in the loss function.
+    gamma : float
+        Weight of the KL divergence term in the loss function.
+    encoder : nn.Sequential
+        Encoder network.
+    fc1 : nn.Linear
+        Fully connected layer 1 in the encoder.
+    fc_mu : nn.Linear
+        Fully connected layer to output the mean of the latent space.
+    fc_logvar : nn.Linear
+        Fully connected layer to output the log variance of the latent space.
+    fc2 : nn.Linear
+        Fully connected layer 2 in the decoder.
+    fc3 : nn.Linear
+        Fully connected layer 3 in the decoder.
+    decoder_input : nn.Linear
+        Input layer for the decoder.
+    decoder : nn.Sequential
+        Decoder network.
+    final_layer : nn.Linear
+        Final layer of the decoder.
+    
+    Methods:
+    --------
+    encode(input: torch.Tensor) -> List[torch.Tensor]
+        Encodes the input into the latent space.
+    decode(z: torch.Tensor) -> torch.Tensor
+        Decodes the latent variable into the reconstructed input.
+    reparameterize(mu: torch.Tensor, logvar: torch.Tensor) -> torch.Tensor
+        Reparameterization trick to sample from the latent space.
+    forward(x: torch.Tensor)
+        Forward pass through the network.
+    log_density_gaussian(x: torch.Tensor, mu: torch.Tensor, logvar: torch.Tensor)
+        Computes the log density of a Gaussian.
+    loss_function(*args, **kwargs) -> dict
+        Computes the loss for the Beta-TCVAE.
+    """
+    
+    num_iter = 0  # Global static variable to keep track of iterations
+    
+    def __init__(self,
+                 input_dim: int,
+                 latent_dim: int,
+                 beta: float,
+                 hidden_dims: List = None,
+                 alpha: float = 1.0,
+                 gamma: float = 1.0,
+                 anneal_steps: int = 200,
+                 **kwargs) -> None:
+        super(BetaTCVAE, self).__init__()
+
+        # Initialize class attributes
+        self.latent_dim = latent_dim
+        self.anneal_steps = anneal_steps
+        self.alpha = alpha
+        self.beta = beta
+        self.gamma = gamma
+
+        if hidden_dims is None:
+            hidden_dims = [256, 256, 256]  # Default hidden dimensions if none are provided
```

Comment:
Consider making this an optional arg with these dimension as the default. This will cut down on lines of code. I will make the suggestion above

Reply:
fixed